### PR TITLE
Space - Katie

### PR DESCRIPTION
**Pull request by ktvoort123**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model in Rails is dealing with the data for the application and interacts with the database.
Describe in your own words what the Controller is doing in Rails | The controller is the central processing of a rails app.
Describe in your own words what the View is doing in Rails | The view takes in data and outputs HTML.
Describe an edge-case controller test you wrote | N/A
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params prevents assigning request parameters to objects unless they have been explicitly permitted.
How are Rails migrations related to Rails models? |  A Migration is a change to a database schema. The Model is a class that is being represented by the database schema.
Describe one area of Rails that are still unclear on | I get the most confused in Rails differentiating what Rails does with its 'magic' and what I have to do.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Rails.application.routes.draw do
```
Comment (CheezItMan) [inline config/routes.rb:1]: 👍

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  <header>
+    <h1>Welcome to Tasks List</h1>
+    <nav>
+      <ul>
+        <li><%= link_to "Create a new task", new_task_path %></li>
+        <li><%= link_to "View all tasks", tasks_path %></li>
+
+      </ul>
+    </nav>
+  </header>
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:21]: I like this, but it should be the `<body>` tag.

File: app/views/tasks/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<%= form_with model: @task, class: 'edit-task' do |f| %>
```
Comment (CheezItMan) [inline app/views/tasks/edit.html.erb:1]: I realize this was reading after Task-List was due, but in future projects you can dry things up like this form with a partial view.  Just a reminder.

File: app/views/tasks/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<h1>Tasks List</h1>
+<ul>
+  <%@tasks.each do |task|%>
+    <li>
+      <%= link_to task.name, task_path(task) %> <%= button_to "Delete Task", task_path(task), method: :delete, data: {confirm: "Are you sure?"}%> <%= button_to "Complete Task", task_complete_path(task), method: :patch%>
```
Comment (CheezItMan) [inline app/views/tasks/index.html.erb:5]: This is a really long line, maybe break this up over multiple lines.

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+
+  def create
+    @task = Task.new(name: params[:task][:name], description: params[:task][:description]) # instantiate a new task
+    
+    if @task.save # save returns true if the database insert succeeds
+      redirect_to task_path(@task.id) 
+      return
+    else
+      render :new, :bad_request # show the new task form view again
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:33]: I just find this more readable, but I think yours is correct too. 

```suggestion
      render :new, status: :bad_request # show the new task form view again
```

File: app/controllers/tasks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    @task = Task.find_by(id: params[:id])
+   
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    elsif @task.update(name: params[:task][:name], description: params[:task][:description]) # update task
+      redirect_to task_path(@task.id)
+      return
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:59]:
```suggestion
      render :edit, status: :bad_request
```

File: app/controllers/tasks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    @task = Task.find_by(id: params[:id])
+   
+    if @task.nil?
+      redirect_to tasks_path
+      return
+    elsif @task.update(completed_at: Time.now) # update task
+      redirect_to task_path(@task.id)
+      return
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/tasks_controller.rb:91]: Why render edit here?  Maybe instead just redirect them someplace.

**Comment by CheezItMan** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️, html is good not much styling
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️, I would suggest making this toggle between complete and incomplete tasks

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅



## Summary


Nicely done, you hit the learning goals here.  You've definitely demonstrated an understanding of the lessons we modeled in Ada Books.  Good luck with Rideshare!
